### PR TITLE
Ring buffer

### PR DESCRIPTION
**Pull request by jonastheis**
This PR adds a simple, fixed capacity ring buffer that replaces elements in FIFO order.

### CODE REVIEW

File: ds/ringbuffer/ringbuffer.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+	r.buffer[r.pos] = element
+	r.pos = (r.pos + 1) % r.capacity
+
+	if r.size < r.capacity {
+		r.size = r.size + 1
+	}
+}
+
+// Get returns all the elements currently in the buffer, from newest to oldest.
+func (r *RingBuffer[T]) Get() []T {
```
Comment (alexsporn) [inline ds/ringbuffer/ringbuffer.go:39]: There is only one possible way to get elements out?

File: ds/ringbuffer/ringbuffer.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+// Add adds an element to the buffer, overwriting the oldest element if the buffer is full.
+func (r *RingBuffer[T]) Add(element T) {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+
+	r.buffer[r.pos] = element
+	r.pos = (r.pos + 1) % r.capacity
+
+	if r.size < r.capacity {
+		r.size = r.size + 1
```
Comment (alexsporn) [inline ds/ringbuffer/ringbuffer.go:34]: The size never decreases? So this can never be emptied again

**Review by piotrm50 — APPROVED** [review]
LGTM, but Alex raised some good questions.

File: ds/ringbuffer/ringbuffer.go (hunk 1 above)
Comment (karimodm) [reply ds/ringbuffer/ringbuffer.go:39]: We wanted to make the API as minimal as possible to our use-case, and not bloat it with methods we didn't gonna use.

File: ds/ringbuffer/ringbuffer.go (hunk 2 above)
Comment (karimodm) [reply ds/ringbuffer/ringbuffer.go:34]: Correct, size cannot decrease. We designed it after a regular buffer: you cannot decrease the size of it: the only way is to assign a new buffer and slice the old one with the element removed.

**Review by alexsporn — APPROVED** [review]

**Pull request merged**